crates/rustc_codegen_spirv/src/codegen_cx/entry.rs

```diff
@@ -92,52 +113,85 @@ impl<'tcx> CodegenCx<'tcx> {
+            } => (return_type, arguments),
             other => self.tcx.sess.fatal(&format!(
                 "Invalid entry_stub type: {}",
                 other.debug(entry_func.ty, self)
             )),
         };
         let mut decoration_locations = HashMap::new();
         // Create OpVariables before OpFunction so they're global instead of local vars.
-        let arguments = entry_fn_abi
-            .args
-            .iter()
-            .zip(hir_params)
-            .map(|(entry_fn_arg, hir_param)| {
-                self.declare_parameter(entry_fn_arg.layout, hir_param, &mut decoration_locations)
-            })
-            .collect::<Vec<_>>();
+        let new_spirv = self.emit_global().version().unwrap() > (1, 3);
+        let arg_len = arg_abis.len();
+        let mut arguments = Vec::with_capacity(arg_len);
+        let mut interface = Vec::with_capacity(arg_len);
+        let mut rta_lens = Vec::with_capacity(arg_len / 2);
+        let mut arg_types = entry_func_arg_types.iter();
+        for (hir_param, arg_abi) in hir_params.iter().zip(arg_abis) {
+            // explicit next because there are two args for scalar pairs, but only one param & abi
+            let arg_t = *arg_types.next().unwrap_or_else(|| {
+                self.tcx.sess.span_fatal(
+                    hir_param.span,
+                    &format!(
+                        "Invalid function arguments: Param {:?} Abi {:?} missing type",
+                        hir_param, arg_abi.layout.ty
+                    ),
+                )
+            });
+            let (argument, storage_class) =
+                self.declare_parameter(arg_abi.layout, hir_param, arg_t, &mut decoration_locations);
+            // SPIR-V <= v1.3 only includes Input and Output in the interface.
+            if new_spirv
+                || storage_class == StorageClass::Input
+                || storage_class == StorageClass::Output
+            {
+                interface.push(argument);
+            }
+            arguments.push(argument);
+            if let SpirvType::Pointer { pointee } = self.lookup_type(arg_t) {
+                if let SpirvType::Adt {
+                    size: None,
+                    field_types,
+                    ..
+                } = self.lookup_type(pointee)
+                {
+                    let len_t = *arg_types.next().unwrap_or_else(|| {
+                        self.tcx.sess.span_fatal(
+                            hir_param.span,
+                            &format!(
+                                "Invalid function arguments: Param {:?} Abi {:?} fat pointer missing length",
+                                hir_param, arg_abi.layout.ty
+                            ),
+                        )
+                    });
+                    rta_lens.push((arguments.len() as u32, len_t, field_types.len() as u32 - 1));
+                    arguments.push(u32::MAX);
+                }
+            }
+        }
```

Comment:
This seems pretty ad-hoc and confusing to follow (splitting it up into functions would probably help), and has some issues with it. For example, this compiletest passes, but it shouldn't (no way to get the length of the slice):

```
// build-pass

use spirv_std::storage_class::{Output, Uniform};

#[spirv(fragment)]
pub fn shader(
    #[spirv(descriptor_set = 0, binding = 0)] slice: Uniform<&[f32]>,
    mut out: Output<f32>,
) {
    let x = slice[5];
    *out = x;
}
```